openmc/dagmc.py

```diff
@@ -0,0 +1,630 @@
+
+        Parameters
+        ----------
+        geom_type : str
+            The type of geometric entity to count. One of {'Volume', 'Surface'}. Returns
+            the runtime number of voumes in the DAGMC model (includes implicit complement).
+
+        Returns
+        -------
+        int
+            Number of geometry elements of the specified type
+        """
+        cv.check_value('geometry type', geom_type, ('volume', 'surface'))
+
+        def decode_str_tag(tag_val):
+            return tag_val.tobytes().decode().replace('\x00', '')
+
+        with h5py.File(self.filename) as dagmc_file:
+            category_data = dagmc_file['tstt/tags/CATEGORY/values']
+            category_strs = map(decode_str_tag, category_data)
+            n = sum([v == geom_type.capitalize() for v in category_strs])
+
+            # check for presence of an implicit complement in the file and
+            # increment the number of cells if it doesn't exist
+            if geom_type == 'volume':
+                name_data = dagmc_file['tstt/tags/NAME/values']
+                name_strs = map(decode_str_tag, name_data)
+                if not sum(['impl_complement' in n for n in name_strs]):
+                    n += 1
+        return n
+
+    @property
+    def n_cells(self):
+        return self._n_geom_elements('volume')
+
+    @property
+    def n_surfaces(self):
+        return self._n_geom_elements('surface')
+
+    def create_xml_subelement(self, xml_element, memo=None):
+        if memo is None:
+            memo = set()
+
+        if self in memo:
+            return
+
+        memo.add(self)
+
+        # Ensure that the material overrides are up-t-date
+        self.build_override_mat_from_cells()
+
+        # Set xml element values
+        dagmc_element = ET.Element('dagmc_universe')
+        dagmc_element.set('id', str(self.id))
+
+        if self.auto_geom_ids:
+            dagmc_element.set('auto_geom_ids', 'true')
+        if self.auto_mat_ids:
+            dagmc_element.set('auto_mat_ids', 'true')
+        dagmc_element.set('filename', str(self.filename))
+        if self._material_overrides:
+            mat_element = ET.Element('material_overrides')
+            for key in self._material_overrides:
+                cell_overrides = ET.Element('cell')
+                cell_overrides.set("id", str(key))
+                cell_overrides.set("material",  ' '.join(
+                    str(t.id) for t in self._material_overrides[key]))
+                mat_element.append(cell_overrides)
+            dagmc_element.append(mat_element)
+        xml_element.append(dagmc_element)
+
+    def build_override_mat_from_cells(self):
+        """
+        Builds the material override dictionary for cells with multiple instances.
+
+        Returns:
+            None
+        """
+        for cell in self.cells.values():
+            if isinstance(cell.fill, Iterable):
+                self.add_material_override(cell, [mat for mat in cell.fill if mat])
```

Comment:
Won't this alter the intended cell fills if some of the entries in the `cell.fill` are skipped?